crates/autopilot/src/run_loop.rs

```diff
@@ -721,40 +721,57 @@ impl RunLoop {
         request: &solve::Request,
     ) -> Result<Vec<Result<competition::Solution, domain::competition::SolutionError>>, SolveError>
     {
+        let authenticator = self.eth.contracts().authenticator();
+        let is_allowed = authenticator
+            .is_solver(driver.submission_address.into())
+            .call()
+            .await;
+
+        // Do not send the request to the driver if the solver is denied
+        match is_allowed {
+            Ok(true) => {}
+            Ok(false) => return Err(SolveError::SolverDenyListed),
+            Err(err) => {
+                tracing::warn!(
+                    driver = driver.name,
+                    ?driver.submission_address,
+                    ?err,
+                    "failed to check if solver is deny listed"
+                );
+                return Err(SolveError::SolverDenyListed);
+            }
+        }
+
         let response = tokio::time::timeout(self.config.solve_deadline, driver.solve(request))
             .await
             .map_err(|_| SolveError::Timeout)?
             .map_err(SolveError::Failure)?;
         if response.solutions.is_empty() {
             return Err(SolveError::NoSolutions);
         }
-        let solutions = response.into_domain();
-
-        // TODO: remove this workaround when implementing #2780
-        // Discard any solutions from solvers that got deny listed in the mean time.
-        let futures = solutions.into_iter().map(|solution| async {
-            let solution = solution?;
-            let solver = solution.solver();
-            let authenticator = self.eth.contracts().authenticator();
-            let is_allowed = authenticator.is_solver(solver.into()).call().await;
-
-            match is_allowed {
-                Ok(true) => Ok(solution),
-                Ok(false) => Err(domain::competition::SolutionError::SolverDenyListed),
-                Err(err) => {
-                    // log warning but discard solution anyway to be on the safe side
-                    tracing::warn!(
-                        driver = driver.name,
-                        ?solver,
-                        ?err,
-                        "failed to check if solver is deny listed"
-                    );
-                    Err(domain::competition::SolutionError::SolverDenyListed)
-                }
-            }
-        });
-
-        Ok(futures::future::join_all(futures).await)
+        // Filter the responses
+        Ok(response
+            .into_domain()
+            .into_iter()
+            .filter(|solution| {
```

Comment:
can we move the filtering to "competition" function? This is where I started putting all the checking/filtering of solutions so everything is in one place. So at the place where the fairness is checked